File: code/class-01-introduction-to-oop/hello_oop.py
"""
hello_oop.py

Class 01 – Introduction to Object-Oriented Programming

Demonstrates:
  - Defining a class with attributes and a method
  - Creating objects (instances)
  - Accessing object state and behaviour
"""


# ── A minimal class ──────────────────────────────────────────────────────────

class Animal:
    """A simple class representing an animal."""

    def __init__(self, name: str, sound: str) -> None:
        # Instance attributes (state)
        self.name  = name
        self.sound = sound

    # Method (behaviour)
    def make_sound(self) -> None:
        print(f"{self.name} says: {self.sound}")


# ── Entry point ───────────────────────────────────────────────────────────────

if __name__ == "__main__":
    # Create two objects from the same class blueprint
    dog = Animal("Rex",      "Woof")
    cat = Animal("Whiskers", "Meow")

    dog.make_sound()   # Rex says: Woof
    cat.make_sound()   # Whiskers says: Meow

    # Each object has its own independent state
    print(f"Dog name : {dog.name}")
    print(f"Cat name : {cat.name}")

File: code/class-01-introduction-to-oop/HelloOOP.java
/**
 * HelloOOP.java
 *
 * Class 01 – Introduction to Object-Oriented Programming
 *
 * Demonstrates the most basic OOP concepts:
 *   - Defining a class with fields and a method
 *   - Creating objects (instances)
 *   - Accessing object state and behaviour
 */
public class HelloOOP {

    // ----- A minimal class -----
    static class Animal {
        // Fields (state)
        String name;
        String sound;

        // Constructor
        Animal(String name, String sound) {
            this.name  = name;
            this.sound = sound;
        }

        // Method (behaviour)
        void makeSound() {
            System.out.println(name + " says: " + sound);
        }
    }

    // ----- Entry point -----
    public static void main(String[] args) {
        // Create two objects from the same class blueprint
        Animal dog = new Animal("Rex",      "Woof");
        Animal cat = new Animal("Whiskers", "Meow");

        dog.makeSound(); // Rex says: Woof
        cat.makeSound(); // Whiskers says: Meow

        // Each object has its own independent state
        System.out.println("Dog name: " + dog.name);
        System.out.println("Cat name: " + cat.name);
    }
}
